ref(grouping): Add `GroupingComponent` subclasses

lobsterkatie opened the pull request:
This adds a full taxonomy of `GroupingCompontent` subclasses, partially just as a form of documentation (we create this deeply nested `variants` structure when grouping, and it's helpful to understand what kind of data is in each part), and partially so that it's easier to write helpers dealing with different kinds of components and not have to constantly be checking for the presence of this or that attribute.

In order to support the subclassing, as few other changes have been made:

- Rename `GroupindComponent` to `BaseGroupingComponent`.
- Add class-level type declarations to `BaseGroupingComponent`.
- Allow `id` to be a class attribute (no longer require it to be passed to `__init__`).
- Use the class name (rather than hardcoded `GroupingComponent`) in `__repr__`.
- Add `int` to the `values` list type. (This is more of a bug fix, though it's also a necessary for the typing of `NSErrorGroupingComponent`. In fact, `values` has always been able to contain ints, but the incoming data isn't super strictly typed, so typechecking has passed even with this missing.)

JoshFerge reviewed (commented):
Adding a bit of color to the above:

Basically what it comes down to is basically I'm advocating for "replace type code with class" https://refactoring.guru/replace-type-code-with-class. while the current implementation is "Replace typecode with subclass https://refactoring.guru/replace-type-code-with-subclasses".

My arguments for "replace type code with class" are 
- Sidestep all brittleness / confusion / complication with python's current type system and our previous setup
- the current implementation does not actually enforce the types its setting
- avoiding a somewhat complex inheritance hierarchy
- use enum to constrain the values of `id`
- can do runtime check using map of Enums / "ALLOW_PRIMITIVES" to enforce correctness, https://github.com/getsentry/sentry/commit/388b1b27255710783a0f8b8da762734f03c71947 as well as documenting it in code
- Given current implementation, we aren't taking advantage of polymorphism, as in https://github.com/getsentry/sentry/pull/80534/ we're still having to do `isinstance` checks, so we aren't gaining a whole lot in my mind

Arguments for "replace type-code with subclass"
- Could eventually in theory validate at type-checking time instead of run-time
- could eventually take advantage of more polymorphic patterns moving things to instance methods, and removing various `isinstance` calls
- in helper functions https://github.com/getsentry/sentry/pull/80534/ it makes some parts easier as we know what's in the values

armenzg reviewed (changes requested):
I see what you're aiming at but as Josh pointed out this is not a good design pattern to follow.

lobsterkatie commented:
So, prompted by Josh, I made a few changes. 

TL;DR, using a generic worked this time, so now all of the subclasses just inherit from `BaseGroupingComponent`, with different types provided for its type parameter. This means types are actually being enforced at the base class level, both in `__init__` and in `.update()`. (In fact, implementing this led me to discover a few existing type issues, which are fixed [here](https://github.com/getsentry/sentry/pull/80682) and [here](https://github.com/getsentry/sentry/pull/80724).)

I also did the following:

- Switched the security strategies so they're just like the any other strategy, in that they just create a component directly, rather than relying on the helper to which I was having to pass the class type. That's [here](https://github.com/getsentry/sentry/pull/80726).
- Pulled the changes from this PR needed by but not directly related to the subclassing out into a separate PR, since they're things which can happen regardless. That's [here](https://github.com/getsentry/sentry/pull/80723).
- Just to reduce the noise, pulled the renaming of `GroupingComponent` to `BaseGroupingComponent` out as well. Given that it really only makes sense if the subclassing is approved, I won't merge it until we're agreed on that. That's [here](https://github.com/getsentry/sentry/pull/80725).
- Because that all took some rearranging of branches, opened a new PR with just the subclassing [here](https://github.com/getsentry/sentry/pull/80722).

Hopefully that addresses the concerns? Please let me know if not!

(Note: Right now all the PRs are stacked, because they're all necessary for the subclass changes, but most of them are independent, so once we're all set I'll rebase them on to `master` separately so their CI can run in parallel and I can get them merged.)

Oh, also: The one remaining typing question I had: I don't love that I have to type `contributing_variant` and `contributing_component` here XXXXX as `Any`. I wanted to be able to associate the correct type for each with the `hash_basis`, but including the types in  `GROUPING_METHODS_BY_DESCRIPTION` didn't work, because you can't pass anything but a literal type to `cast`. Definitely open to suggestions if there's a nice way to make that work. (UPDATE, as I'm typing this: Overloads? Maybe overloads. Lemme try that.)

JoshFerge commented on src/sentry/grouping/component.py:
```diff
@@ -23,33 +23,38 @@
 
 
-class GroupingComponent:
+class BaseGroupingComponent:
```
can this be made abstract?

JoshFerge commented on src/sentry/grouping/component.py:
```diff
@@ -23,33 +23,38 @@
     def __init__(
         self,
-        id: str,
+        id: str | None = None,
```
why would we ever pass in id if there are specific subclasses for each grouping component?

JoshFerge commented:
@lobsterkatie i don't think the changes you're mentioning in the above comment are pushed up?

lobsterkatie commented:
> @lobsterkatie i don't think the changes you're mentioning in the above comment are pushed up?

Sorry, maybe this wasn't clear:

> Because that all took some rearranging of branches, I opened a new PR with just the subclassing [here](https://github.com/getsentry/sentry/pull/80722).

It just seemed easier, given that I breaking things up, to make a new PR stack. Let me close this one to make it clearer. (I'll also answer your other questions there.)
